lewis/adapters/modbus.py

```diff
@@ -0,0 +1,450 @@
+        :return: ModbusTCPFrame instance that represents an exception
+        """
+        frame = deepcopy(self)
+        frame.length = 3
+        frame.fcode += 0x80
+        frame.data = bytearray(chr(code))
+        return frame
+
+    def create_response(self, data=None):
+        """
+        Create a response frame based on this frame.
+
+        :param data: Data section of response as bytearray. If None, request data section is kept.
+        :return: ModbusTCPFrame instance that represents a response
+        """
+        frame = deepcopy(self)
+        if data is not None:
+            frame.data = data
+        frame.length = 2 + len(frame.data)
+        return frame
+
+
+class ModbusProtocol(object):
+    """
+    StateMachine based Modbus Protocol parser.
+    """
+
+    def _init_state(self):
+        self._request = None
+        self._response = None
+        self._exception = None
+
+    def __init__(self, handler, databank):
+        self._init_state()
+
+        self.handler = handler
+        self.databank = databank
+        self.buffer = bytearray()
+
+        self.protocol = StateMachine({
+            'initial': 'await_request',
+            'transitions': OrderedDict([
+                (('await_request', 'check_fcode'), lambda: self._request is not None),
+                (('check_fcode', 'send_response'), lambda: self._exception is not None),
+
+                (('check_fcode', 'read_bit'),
+                 lambda: self._request.fcode in
+                    (MBFC.READ_COILS, MBFC.READ_DISCRETE_INPUTS)),
+                (('check_fcode', 'read_word'),
+                 lambda: self._request.fcode in
+                    (MBFC.READ_HOLDING_REGISTERS, MBFC.READ_INPUT_REGISTERS)),
+                (('check_fcode', 'write_coil'),
+                 lambda: self._request.fcode == MBFC.WRITE_SINGLE_COIL),
+                (('check_fcode', 'write_register'),
+                 lambda: self._request.fcode == MBFC.WRITE_SINGLE_REGISTER),
+
+                (('read_bit', 'send_response'), lambda: True),
+                (('read_word', 'send_response'), lambda: True),
+                (('write_coil', 'send_response'), lambda: True),
+                (('write_register', 'send_response'), lambda: True),
+                (('send_response', 'await_request'), lambda: True),
+
+                # Print unknown packets
+                (('check_fcode', 'print_packet'), lambda: True),
+                (('print_packet', 'await_request'), lambda: True),
+            ])
+        })
+        self.protocol.bind_handlers_by_name(self, prefix=['_on_', '_in_', '_after_'])
+        self.protocol.process()  # Cycle into initial state
+
+    def parse(self, data=None):
+        """
+        Parse as much of current buffer as possible.
+
+        :param data: Optionally append given data to buffer prior to parsing
+        """
+        self.buffer.extend(bytearray(data or ''))
+
+        # Process until two or three FSM cycles stay in the same state
```

Comment:
I'm not sure if this applies to Modbus over TCP, but in general there is no guarantee in the protocol that requests arrive as one unit. When I started, I was considering using StateMachine to parse the incoming stream byte by byte (or word by word). 

But in the end, it seemed far simpler to encapsulate reading in ModbusTCPFrame and just use ModbusProtocol and its StateMachine to step through the higher level protocol exchange. Not entirely sure using StateMachine is even of advantage now... the only real branching point is after checking the function code, and that could be done with a function table.